CryptomatorFileProvider/Locks/LockManager.swift

```diff
@@ -38,14 +38,16 @@ import Promises
  ```
  */
 class LockManager {
-	private static var pathLocks = [String: RWLock]()
-	private static var dataLocks = [String: RWLock]()
-	private static let queue = DispatchQueue(label: "LockManager Queue", qos: .userInitiated, attributes: .concurrent)
-	private static let dictionaryQueue = DispatchQueue(label: "LockManager dicitionaryQueue")
-
-	private static func readLock(locks: [RWLock], paths: [CloudPath]) -> Promise<LockNode> {
+	static let cloud = LockManager()
+	static let local = LockManager()
+	private var pathLocks = MapTable<NSString, RWLock>(keyOptions: .copyIn, valueOptions: .weakMemory) // [String: RWLock]()
+	private var dataLocks = MapTable<NSString, RWLock>(keyOptions: .copyIn, valueOptions: .weakMemory) // [String: RWLock]()
```

Comment:
The idea was as follows: we always have a reference to the locks until the task is completed and we release (unlock) the locks. If a new task is added during this time, it can use (and wait for) the existing locks since they are still in the MapTable. As soon as nobody uses a certain lock anymore, i.e. it has already been unlocked and no other task is waiting to acquire it, it gets removed from the MapTable and the memory. The reason for this is that otherwise we would permanently keep the locks for all paths ever visited (in a session) in this dictionary and thus use up memory unnecessarily. But I agree with you that this might be over-optimization (which could lead to unexpected behavior), since a single RWLock instance doesn't use much memory.